operator/rediscontroller/pipeline.go

```diff
@@ -0,0 +1,226 @@
+package rediscontroller
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	exoscalesdk "github.com/exoscale/egoscale/v2"
+	"github.com/exoscale/egoscale/v2/api"
+	"github.com/exoscale/egoscale/v2/oapi"
+	"github.com/go-logr/logr"
+	exoscalev1 "github.com/vshn/provider-exoscale/apis/exoscale/v1"
+	"github.com/vshn/provider-exoscale/operator/mapper"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// pipeline is a managed.ExternalClient and implements a crossplane reconciler for redis.
+type pipeline struct {
+	kube     client.Client
+	recorder event.Recorder
+	exo      *exoscalesdk.Client
+}
+
+// newPipeline returns a new instance of pipeline.
+func newPipeline(client client.Client, recorder event.Recorder, exoscaleClient *exoscalesdk.Client) *pipeline {
+	return &pipeline{
+		kube:     client,
+		recorder: recorder,
+		exo:      exoscaleClient,
+	}
+}
+
+func (p pipeline) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
```

Comment:
I like this approach of keeping the methods inside one file though at this point I would prefer we keep the structure consistent that is a method per file. We can always decide to do some refactoring later on.

Reply:
ok, fair enough.